socrata/socrata_upload.py

```diff
@@ -0,0 +1,337 @@
+        groups = []
+        for i in range(len(years)):
+            for j in range(len(township_codes)):
+                groups.append((years[i], township_codes[j]))
+
+        flag = "both"
+
+    else:
+        if not years:
+            groups = None
+            flag = None
+        else:
+            groups = years
+            flag = "years"
+
+    return flag, groups
+
+
+def socrata_upload(
+    socrata_asset, overwrite=False, years=None, by_township=False
+):
+    """
+    Wrapper function for building SQL query, retrieving data from Athena, and
+    uploading it to Socrata. Allows users to specify target Athena and Socrata
+    assets, define columns to construct a `row_id`, whether the data on Socrata
+    should be overwritten or updated, and whether or not to chunk the upload by
+    year. By default the function will query a given Athena asset by year for
+    all years and upload via `post` (update rather than overwrite).
+    """
+
+    athena_asset, asset_id, row_identifier = get_asset_info(socrata_asset)
+
+    flag, groups = generate_groups(
+        years=years, by_township=by_township, athena_asset=athena_asset
+    )
+
+    tic = time.perf_counter()
+    count = 0
+
+    if not flag:
+        sql_query = build_query(
+            athena_asset=athena_asset,
+            row_identifier=row_identifier,
+        )
+
+        upload_args = {
+            "asset_id": asset_id,
+            "sql_query": sql_query,
+            "overwrite": overwrite,
+        }
+
+        if overwrite:
+            response = upload("put", **upload_args)
+        else:
+            response = upload("post", **upload_args)
+        print(response.content)
+
+    else:
+        if flag == "years":
+            sql_query = build_query(
+                athena_asset=athena_asset,
+                row_identifier=row_identifier,
+                years=years,
+            )
+
+        if flag == "both":
+            sql_query = build_query(
+                athena_asset=athena_asset,
+                row_identifier=row_identifier,
+                years=years,
+                township=by_township,
+            )
+
+        if overwrite:
+            for item in groups:
+                if flag == "both":
+                    upload_args = {
+                        "asset_id": "asset_id",
+                        "sql_query": "sql_query",
+                        "overwrite": "overwrite",
+                        "year": item[0],
+                        "township": item[1],
+                    }
+                elif flag == "years":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item,
+                    }
+                if count == 0:
+                    response = upload("put", **upload_args)
+                else:
+                    response = upload("post", **upload_args)
```

Comment:
**[Question, non-blocking]** This is somewhat counterintuitive to me, since that my understanding of the distinction between `PUT` vs. `POST` is that `POST` is typically used to create a resource and `PUT` is used to update it (see [this SO question](https://stackoverflow.com/q/630453) for links to the RFC definitions). Wouldn't we expect to `POST` on the first request, and `PUT` thereafter?

Reply:
For whatever reason, here's [how socrata does it](https://dev.socrata.com/publishers/soda-producer/soda-producer-basics).